stumpy/scrump.py

```diff
@@ -112,58 +118,85 @@ def _compute_PI(
             squared_distance_profile[zone_start : zone_stop + 1] = np.inf
 
             # only for self-join
-            mask = squared_distance_profile < P_squared[thread_idx]
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
-
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
+            IDX = np.flatnonzero(
+                squared_distance_profile < P_squared[thread_idx, :, -1]
+            )
+            for idx in IDX:
+                pos = np.searchsorted(
+                    P_squared[thread_idx, idx],
+                    squared_distance_profile[idx],
+                    side="right",
+                )
+                core._insert(
+                    P_squared[thread_idx, idx, :], pos, squared_distance_profile[idx]
+                )
+                core._insert(I[thread_idx, idx, :], pos, i)
+
+        idx = np.argmin(squared_distance_profile)
+        core._insert(P_squared[thread_idx, i, :], 0, squared_distance_profile[idx])
```

Comment:
**In below, I am trying to explain the concept of adding `for idx in IDX` for  `self-join` (then, I will explain `top-k`)**

The `for idx in IDX` loop tries to update MOST OF (not ALL OF) matrix profile with the (squared) distance profile  `squared_distance_profile` calculated for one of the  subsequences (subsequence `S_i = T[i: i+m]`). So, we UPDATE `matrix_profile[index]` by `squared_distance_profile[index]`. The reason that I used the term `MOST OF` earlier is that the squared_distance_profile[i] and its trivial matches are set to `inf`. So, we are not updating matrix profile for indices in range `(i - excl_zone, i + excl_zone)`. 

The `matrix profile[i]` is EXACTLY the minimum of the `squared_distance_profile` calculated for `S_i`. That is why the original paper has two parts: one for UPDATEing (self-join) matrix profile and one for SETting the `matrix profile[i]` by finding the minimum of  distance profile.

Now, in top-k, we want to keep the main structure of the algorithm. So, when we want to UPDATE `TopK matrix profile` by using the `S_i`'s distance profile (`squared_distance_profile`),  we compare the new value with the k-th value (largest value in  `TopK matrix profile`). Again, note that we are doing it for `out-of-trivial-zone` indices. Now, at the index `i` itself, we know that minimum of distance  profile is EXACTLY the value at index  0 (smallest distance of `S_i` to its neighbors)

---

**Okay, now I can see that I could have SET `TopK matrix profile[i]` like this:**
```
# instead of np.argmin()...and just SET 0-th value of `TopK matrix profile[i]`, we can SET all k values for `S_i`
# because, we have the distance profile of `S_i` and that is all I need for getting TopK matrix profile values for `S_i`

IDX = np.argsort(squared_distance_profile)[:k] # or np.argpartition and then sort
P_squared[thread_idx, i, :] = IDX 
```

Reply:
> What do you think?

I totally understand. I will try to come up with a brief yet nice comment to explain this part. I can also try to see if I can revise it (choosing structure? choosing better  name? ...). I will try to make it better in the next set of commits.

Reply:
Thank you